sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1147,7 +1147,7 @@ object CollapseProject extends Rule[LogicalPlan] with AliasHelper {
   /**
    * Check if the given expression is cheap that we can inline it.
    */
-  private def isCheap(e: Expression): Boolean = e match {
+  def isCheap(e: Expression): Boolean = e match {
```

Comment:
Note that `isCheap` returns true for `PythonUDF`. It's OK for `CollapseProject` as the python udf planner will deduplicate python udfs in a single project operator. I'm not sure about multi-like optimization though.

Reply:
nvm, we split multi-like into multiple string predicates and they are still in the same operator. python udf should be fine.